kani-compiler/src/kani_middle/attributes.rs

```diff
@@ -83,10 +113,60 @@ impl<'tcx> KaniAttributes<'tcx> {
         Self { map, tcx, item: def_id }
     }
 
+    /// Expect that at most one attribute of this kind exists on the function
+    /// and return it.
+    fn expect_maybe_one(&self, kind: KaniAttributeKind) -> Option<&'tcx Attribute> {
+        match self.map.get(&kind)?.as_slice() {
+            [one] => Some(one),
+            _ => {
+                self.tcx.sess.err(format!(
+                    "Too many {} attributes on {}, expected 0 or 1",
+                    kind.as_ref(),
+                    self.tcx.def_path_debug_str(self.item)
+                ));
+                None
+            }
+        }
+    }
+
+    /// Parse and extract the `proof_for_contract(TARGET)` attribute. The
+    /// returned symbol and defid are respectively the name and id of `TARGET`,
+    /// the span in the span for the attribute (contents).
+    pub fn for_contract(&self) -> Option<(Symbol, DefId, Span)> {
+        self.expect_maybe_one(KaniAttributeKind::ProofForContract).and_then(|target| {
+            let name = expect_key_string_value(self.tcx.sess, target);
+            let resolved = resolve_fn(
+                self.tcx,
+                self.tcx.parent_module_from_def_id(self.item.expect_local()),
+                name.as_str(),
+            );
+            match resolved {
+                Err(e) => {
+                    self.tcx.sess.span_err(
+                        target.span,
+                        format!(
+                            "Failed to resolve replacement function {} because {e}",
+                            name.as_str()
+                        ),
+                    );
+                    None
+                }
+                Ok(ok) => Some((name, ok, target.span)),
+            }
+        })
+    }
+
+    /// Extact the name of the sibling function this contract is checked with
+    /// (if any)
+    pub fn checked_with(&self) -> Option<Symbol> {
+        self.expect_maybe_one(KaniAttributeKind::CheckedWith)
+            .map(|target| expect_key_string_value(self.tcx.sess, target))
+    }
+
     /// Check that all attributes assigned to an item is valid.
     /// Errors will be added to the session. Invoke self.tcx.sess.abort_if_errors() to terminate
     /// the session and emit all errors found.
-    pub(super) fn check_attributes(&self) {
+    pub(super) fn check_attributes(&self, queries: &QueryDb) {
```

Comment:
I was wondering if this check should be applied to `check_unstable_features` instead.